tests/test_modules.hpp

```diff
@@ -0,0 +1,275 @@
+{
+  public:
+    SimpleModule(std::string module_name);
+    SimpleModule(std::string module_name, nlohmann::json config);
+
+    void initialize(segment::Builder& builder) override;
+
+    bool m_was_configured{false};
+
+  private:
+    bool m_initialized{false};
+};
+
+SimpleModule::SimpleModule(std::string module_name) : SegmentModule(std::move(module_name)) {}
+
+SimpleModule::SimpleModule(std::string module_name, nlohmann::json config) :
+  SegmentModule(std::move(module_name), std::move(config))
+{}
+
+void SimpleModule::initialize(segment::Builder& builder)
+{
+    VLOG(10) << "MyModule::operator() called for '" << this->name() << "'" << std::endl;
+
+    if (config().contains("simple_key_1"))
+    {
+        m_was_configured = true;
+    }
+
+    /** First linear path **/
+    auto input1 = builder.make_node<bool, unsigned int>(this->get_module_component_name("input1"),
+                                                        rxcpp::operators::map([this](bool input) {
+                                                            unsigned int output = 42;
+                                                            return output;
+                                                        }));
+
+    auto internal1 = builder.make_node<unsigned int, std::string>(
+        this->get_module_component_name("_internal1_"), rxcpp::operators::map([this](unsigned int input) {
+            auto output = std::to_string(input);
+            VLOG(10) << "Created output1 << " << output << std::endl;
+            return output;
+        }));
+
+    builder.make_edge(input1, internal1);
+
+    auto output1 = builder.make_node<std::string, std::string>(
+        this->get_module_component_name("output1"), rxcpp::operators::map([this](std::string input) { return input; }));
+
+    builder.make_edge(internal1, output1);
+
+    /** Second linear path **/
+    auto input2 = builder.make_node<bool, unsigned int>(this->get_module_component_name("input2"),
+                                                        rxcpp::operators::map([this](bool input) {
+                                                            unsigned int output = 42;
+                                                            return output;
+                                                        }));
+
+    auto internal2 = builder.make_node<unsigned int, std::string>(
+        this->get_module_component_name("_internal2_"), rxcpp::operators::map([this](unsigned int input) {
+            auto output = std::to_string(input);
+            VLOG(10) << "Created output2: " << output << std::endl;
+            return output;
+        }));
+
+    builder.make_edge(input2, internal2);
+
+    auto output2 = builder.make_node<std::string, std::string>(
+        this->get_module_component_name("output2"), rxcpp::operators::map([this](std::string input) { return input; }));
+
+    builder.make_edge(internal2, output2);
+
+    register_input_port("input1", input1, &typeid(bool));
```

Comment:
shouldn't we be able to get the type from the `input1` object?

Reply:
Yeah, this is probably true, I'll clean it up before the final iteration.